src/core/collection.ml

```diff
@@ -0,0 +1,16 @@
+
+module type VectorLike = sig 
```

Comment:
Not a big fan of this name, or this interface. It doesn't seem very well defined. It seems kind of like an `Iterable`, but an `Iterable` would not have a property-like `length` function, but instead use a verb like `count` to imply that an action needs to be performed. It also should not have an `isEmpty` function, but perhaps instead an `any` function (which is probably a better name for `exists` if the predicate is optional)